src/librustc_typeck/check/writeback.rs

```diff
@@ -355,19 +355,18 @@ impl<'cx, 'gcx, 'tcx> WritebackCx<'cx, 'gcx, 'tcx> {
         }
     }
 
-    fn visit_cast_types(&mut self) {
+    fn visit_coercion_casts(&mut self) {
         let fcx_tables = self.fcx.tables.borrow();
-        let fcx_cast_kinds = fcx_tables.cast_kinds();
+        let fcx_coercion_casts = fcx_tables.coercion_casts();
         debug_assert_eq!(fcx_tables.local_id_root, self.tables.local_id_root);
-        let mut self_cast_kinds = self.tables.cast_kinds_mut();
         let common_local_id_root = fcx_tables.local_id_root.unwrap();
 
-        for (&local_id, &cast_kind) in fcx_cast_kinds.iter() {
+        for &local_id in fcx_coercion_casts.iter() {
             let hir_id = hir::HirId {
                 owner: common_local_id_root.index,
                 local_id,
             };
-            self_cast_kinds.insert(hir_id, cast_kind);
+            self.tables.set_coercion_cast(&hir_id);
```

Comment:
Since you are throwing away the HirId during the method call, building a new HirId seems useless. I'm not sure if the owner field ever differs, but I don't think so. Can you investigate what kind of things were checked on the HirId in the `LocalTableInContextMut::insert` method? I don't actually know why (before this PR) we used that type. Also an interesting question is "why are we copying fcx.tables fields manually to self.tables and don't just move the entire tables over?". If you have any trouble figuring out the answers, let me know and I'll see if I can find it out. This seems like some questions whose answer should be documented but might have been forgotten to get documented.

Reply:
> Can you investigate what kind of things were checked on the HirId in the LocalTableInContextMut::insert method?

There is a check to see if the `TypeckTables::local_id_root` and the `HirID::owner` are the same . 
https://github.com/rust-lang/rust/blob/867af819b7f4d45adad4c9dadcbeb3b1e29d2b99/src/librustc/ty/context.rs#L240-L267

I don't understand why is this check there though. It seems redundant to me. This is how cast_kinds is constructed:
https://github.com/rust-lang/rust/blob/a9da8fc9c267c08cfdb8cf5b39da14f154d12939/src/librustc/ty/context.rs#L727-L733
The `LocalTableInContextMut::local_id_root` is always `TypeckTables::local_id_root`.

I guess we can safely remove this check?

Reply:
> Can you investigate what kind of things were checked on the HirId in the LocalTableInContextMut::insert method?

There is a check to see if the `TypeckTables::local_id_root` and the `HirID::owner` are the same . I guess it makes sense because you do not want to insert a` HiriD::local_id` which belongs to a different `local_id_root`?
https://github.com/rust-lang/rust/blob/867af819b7f4d45adad4c9dadcbeb3b1e29d2b99/src/librustc/ty/context.rs#L240-L267

Reply:
Since we are creating the `HirId` with the correct owner, the check is pretty useless for this situation (especially since there's a debug assertion present that already checks whether the owners match up). So... change `set_coercion_cast` to take a `LocalId` and remove the `HirId` creation.

Reply:
But the debug assertion only works if we use a `LocalTableInContext`. We are using just an `ItemLocalSet` for `coercion_casts`

Reply:
Ah, sorry, that was not explained well. There's a debug assertion in https://github.com/rust-lang/rust/pull/58670/files#diff-47724107e5e6f3e247b4463df1f71015L361 (so before the call to `set_coercion_cast`)